### PR TITLE
Modernized front page scenario

### PR DESCRIPTION
**Pull request by philiprbrenan**
---

Used the dag and task decorators to document a more compelling front page which makes Airflow's strengths more immediately visible to potential users.

**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by dirrao — APPROVED** [review]

File: docs/apache-airflow/index.rst [hunk 1]
```diff
@@ -19,56 +19,71 @@ What is Airflow™?
...
 - **Extensible**: The Airflow™ framework contains operators to connect with numerous technologies. All Airflow components are extensible to easily adjust to your environment.
-- **Flexible**: Workflow parameterization is built-in leveraging the `Jinja <https://jinja.palletsprojects.com>`_ templating engine.
+- **Flexible**: Workflows can be easily parameterized by using the built-in `Jinja <https://jinja.palletsprojects.com>`_ templating engine.
 
 Take a look at the following snippet of code:
 
 .. code-block:: python
 
-    from datetime import datetime
+  # Create tasks using a decorated DAG and pass values between those tasks.
```
Comment (potiuk) [inline docs/apache-airflow/index.rst]: I like the proposal in general, but I have one big ask:

I think we should not make the mapped-task example as the first thing the user sees. This is really a bit more "Advanced" use case and not one that ss the main strength of airflow, also it introduces quite a few concepts at once and they might be difficult to grasp by a new user.

Immediate questions:

* shoudl I always do that even in a simple case ?
* what is that expand thing? 
* what the hell those values are.

I think those **maybe** should folllow the basic example as a more advance case (which I would find very useful to add). This should be clearly marked with something like  "while simple workflows are simple, you can also do much more complex tasks where - for example - you can replicate tasks to run in parallel, allowing airflow fo run workflow with semantics similar to MapReduce" - followed by the new example. That would be more appealing to the users who hit the page and have no idea about Airflow.

This should be similar in concept to the "architecture overview" - where we start with a basic, simplest deployment of Airflow and then follow that with more complex cases that focus on distribution and eventually security-focused deployments. 

See https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/overview.html#architecture-diagrams

**Review by potiuk — CHANGES_REQUESTED** [review]
requesting changes so that it's not merged accidentally before addressing the comment about "simple" example

**Comment by philiprbrenan** [comment]
Simplified the initial demo example by replacing the call to ``expand`` with explicit tasks.  I tried using ``>>`` to organize @task decorated methods but this is not permitted requiring the use of parameters instead.  This means that the old way of doing things and the new decorated way of doing things cannot be mixed. The new decorated technique is more much more compelling because it is compact and thus easy to understand, reason about and build upon.

**Review by potiuk — APPROVED** [review]
This is sooo much better

**Review by potiuk — CHANGES_REQUESTED** [review]
Static checks and spelling failures to fix.. 

Also I rendered it locally and I think it needs some more "pragmatic" improvements. This is the main page people land on and it should be very carefully prepared - with the mind of "how users will look at it.

Example should have one heading. Right now it just blends together with previous chapter:

image

It's rather difficult to see and grasp the example because it is not compact enough. The whole example - even in a huge screens spans through more than one screen and it's difficult to see the graph and example together.  See the screenshots below. I think we should make it far more compact and "easier to see" together.

There is a LOT of empty space in the example, maybe you can put the code and image side-by-side in a table (not sure how to do it with sphinx but it should be possible)

image


Also - I think (after deliberation) that we shoudl absolutely not remove the classic operator usage - this is a valid way people are, and will be develioping Airflow DAGs and replacing and even promoting one over the other is not a good idea. 

I think what we need here is to have two examples- classic and task flow that are 1-1 the same. And we already have a very good way of doing it - see https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/python.html#howto-operator-pythonoperator - where we have @task vs. PythonOperator tabs and you can switch between them.

The log image is somewhat strange with the colors

Screenshot from 2024-02-23 14-47-47


So in short: 

* put example to a separate section

* make the example more compact and maybe do something with the screenshots to get them fit into smaller space so that they can be seen together on a single page

* have two swappable variants of the DAG - classic and task flow one

This would be much better.

**Comment by philiprbrenan** [comment]
_put example to a separate  section_

 There are now two examples - a very simple one with **>>** operators, a more complex one with decorators , both in their own sections.

 _make the example more compact and maybe do something with the screenshots to get them fit into smaller space so that they can be seen together on a single page_

Screenshots have been tightened up and reduced by 50%

 _have two swappable variants of the DAG - classic and task flow one_

There are now two examples _classic_:  a very simple one with >> operators,  task with  a more complex one with decorators. A screen shot has been added to show how to run the dags and the log results.

**Comment by potiuk** [comment]
cc: @BasPH - I know you have an experience and are interested in reviewing and "thinking" in the "first contact" kind of docs - I've done my share of comments and directions - curious about your review/comments. This is a super-important part of our documentation so I think we shoudl have at least few poeple who know/care about it to give their feedback.

**Comment by BasPH** [comment]
@philiprbrenan could you elaborate on how you see this change making Airflow's strengths more visible?

One thing I do suggest adding, as @potiuk mentioned, is both ways of writing DAGs (using tabs): (1) the object-oriented approach and (2) the functional approach.

**Comment by potiuk** [comment]
Looks like Chat-GPT generated response. No time to read it, so I shortened it with Chat GPT:

```
Initially skeptical of Airflow, I found it cumbersome and lacking conceptual integrity, akin to scattered building blocks without a coherent design. However, a revelation came when I unearthed a pure Python solution buried deep within Airflow's documentation, utilizing only the @dag and @task decorators. This discovery transformed my perception, turning Airflow from an annoyance to an indispensable tool. Motivated by this experience, I crafted a new introductory approach, emphasizing the simplicity and power of the decorators. Despite suggestions to include the >> operator, I found it confounding and incongruent with typical programming flow. Presenting both approaches risks confusing new users with unnecessary choices, particularly through a GUI interface. Therefore, I advocate for a streamlined introduction focused solely on the decorators, aligning with Airflow's architectural elegance and ensuring a smoother entry for newcomers without unnecessary complexity.
```

This has been discussed multiple times and your point of view is only of yours. Airflow has multiple different kind of users - both experienced Pythonistas and people who are just data scientists and don't care too much about python - they want to put their DAGs from existing building blocks (i.e. operators) 

That's why - what we are asking you to make the first page in the way that appeal to both kind of users - and the tab switching of different ways of programming with Airflow is a good idea. You have to accept the fact of life that other people have different experiences and think differently - and Airflow does not have to respond to your mindset only, but also to other mindsets and our direction guiding you there make it so. Even if not perfectly fitting either, making it appealing to both is our goal.

Kind request to adjust to those expectations.
